crates/oxc_linter/src/lint_adapter.rs

```diff
@@ -0,0 +1,1183 @@
+        JSXOpeningElement, ModuleDeclaration, ObjectPattern, ObjectPropertyKind, PropertyKey,
+        SimpleAssignmentTarget, StringLiteral, TSLiteral, TSSignature, TSType, TSTypeAnnotation,
+        TSTypeName, TemplateLiteral, VariableDeclarator,
+    },
+    AstKind,
+};
+use oxc_semantic::{AstNode, Semantic};
+use oxc_span::{Atom, GetSpan, Span};
+use oxc_syntax::operator::BinaryOperator;
+use serde::Deserialize;
+use trustfall::{
+    provider::{
+        resolve_coercion_with, resolve_neighbors_with, resolve_property_with, Adapter,
+        ContextIterator, ContextOutcomeIterator, EdgeParameters, ResolveEdgeInfo, ResolveInfo,
+        TrustfallEnumVertex, VertexIterator,
+    },
+    FieldValue,
+};
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct InputQuery {
+    pub query: String,
+    pub args: BTreeMap<Arc<str>, FieldValue>,
+    pub reason: String,
+}
+
+#[derive(Debug, Clone, TrustfallEnumVertex)]
+pub enum Vertex<'a> {
+    AstNode(AstNode<'a>),
+    Expression(&'a Expression<'a>),
+    Argument(&'a Argument<'a>, i32), // i32 is argument index
+    Span(Span),
+    Parameter(&'a BindingPatternKind<'a>, i32), // i32 is argument index
+    PossiblyNumber(f64),
+    PropertyAccess(Span, &'a str),
+    PropertyKey(&'a PropertyKey<'a>),
+    TSType(&'a TSType<'a>),
+    TSSignature(&'a TSSignature<'a>),
+    TSTypeAnnotation(&'a TSTypeAnnotation<'a>),
+    TSLiteral(&'a TSLiteral<'a>, &'a Span),
+    TSTypeName(&'a TSTypeName<'a>, &'a Span),
+    EffectivelyLiteralString(&'a Span, &'a Atom),
+    ImportSpecifier(&'a ImportDeclarationSpecifier),
+    JsxElementName(&'a JSXElementName<'a>),
+    JsxOpeningElement(&'a JSXOpeningElement<'a>),
+    JsxClosingElement(&'a JSXClosingElement<'a>),
+    ObjectProperty(&'a ObjectPropertyKind<'a>),
+    VariableDeclaration(&'a VariableDeclarator<'a>),
+    // operators
+    Operator(&'a BinaryOperator),
+    AssignmentOperator(&'a AssignmentExpression<'a>),
+    // assignment to ...
+    AssignmentToIdentifier(&'a Span, &'a Atom),
+    AssignmentOnObject(&'a SimpleAssignmentTarget<'a>),
+    DestructuringAssignment(DestructuringAssignment<'a>),
+
+    File(Rc<AstNodeIterator<'a>>),
+}
+
+#[derive(Debug, Clone)]
+pub enum DestructuringAssignment<'a> {
+    AssignmentExpression(&'a AssignmentTargetPattern<'a>),
+    ArrayPattern(&'a ArrayPattern<'a>),
+    ObjectPattern(&'a ObjectPattern<'a>),
+    DefaultedDestructuringAssignment(&'a AssignmentPattern<'a>),
+}
+
+impl<'a> GetSpan for DestructuringAssignment<'a> {
+    fn span(&self) -> Span {
+        match self {
+            DestructuringAssignment::AssignmentExpression(expr) => expr.span(),
+            DestructuringAssignment::ArrayPattern(expr) => expr.span,
+            DestructuringAssignment::ObjectPattern(expr) => expr.span,
+            DestructuringAssignment::DefaultedDestructuringAssignment(expr) => expr.span,
+        }
+    }
+}
+
+pub struct AstNodeIterator<'a> {
+    iter: RefCell<Option<VertexIterator<'a, Vertex<'a>>>>,
+}
```

Comment:
I have a great feeling that this should be 

```
pub struct AstNodeIterator<'ast, 'strustfall> {
    iter: RefCell<Option<VertexIterator<'trustfall, Vertex<'ast>>>>,
}
```

But I'm afraid of the change and its impact 😭

Reply:
What impact are you afraid of? Best thing you can do is give it a try, at worst it doesn't work and you revert it :)

Reply:
This seems fine to me. Trustfall doesn't carry lifetime parameters of its own, the only lifetime parameter it has is "how long does the underlying data live."

`VertexIterator<'a, T>` is a type alias for `Box<dyn Iterator<Item=T> + 'a>` where `T: 'a`.

So in `VertexIterator<'trustfall, Vertex<'ast>>`, `'ast : 'trustfall` and this would only shorten the lifetime during which a query's results can be retrieved. But we already have the data for `'ast` so I don't see a downside to allowing query results to execute as late as `'ast`.

Reply:
@obi1kenobi My previous message is for the following error:

```rust
error: lifetime may not live long enough
   --> crates/oxc_linter/src/lint_adapter.rs:221:9
    |
218 | impl<'a> LintAdapter<'a> {
    |      -- lifetime `'a` defined here
219 |     fn get_ast_node_iter(&self) -> VertexIterator<'a, Vertex<'a>> {
    |                          - let's call the lifetime of this reference `'1`
220 |         // TODO: Stop cloning this and instead just use iter. Then, remove the pub from nodes
221 |         Box::new(self.semantic.nodes().iter().map(|v| Vertex::AstNode(v)))
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ method was supposed to return data with lifetime `'a` but it is returning data with lifetime `'1`

error: could not compile `oxc_linter` (lib) due to previous error
```

When we tried to remove the clone from 
```rust
impl<'a> LintAdapter<'a> {
    fn get_ast_node_iter(&self) -> VertexIterator<'a, Vertex<'a>> {
        // TODO: Stop cloning this and instead just use iter. Then, remove the pub from nodes
        Box::new(self.semantic.nodes().nodes.clone().into_iter().map(|v| Vertex::AstNode(v)))
    }
}
```

Reply:
See #502 for one possible solution to this.

Reply:
This has been merged!